WIP

thekenu opened the pull request:
### Summary
<!-- Quick summary of changes, optional -->

### Changelist 
<!-- Give a list of the changes covered in this PR. This will help both you and the reviewer keep this PR within scope. -->

### Testing Done
<!-- Outline the testing that was done to demonstrate the changes are solid. This could be unit tests, integration tests, testing on the car, etc. Include relevant code snippets, screenshots, etc as needed. -->

### Resolved Issues
<!-- Link any issues that this PR resolved like so: `Resolves #1, #2, and #5` (Note: Using this format, Github will automatically close the issue(s) when this PR is merged in). -->

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [ ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [ ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

thekenu commented on boards/DIM/Src/App/App_ErrorTable.c:
```diff
@@ -0,0 +1,216 @@
+#include <stdlib.h>
```
`{App|Io}_ErrorTable.{c|h}` are meant to be auto-generated and placed inside `cmake-build-{embedded|gtest}`. I hand-wrote it for now as a prototype.

garethellis0 replied:
Why are there per-board error tables but also the shared one?

thekenu replied:
The intention was to have only one error table shared across the whole car. I think it became confusing when i decided to put the code in `DIM` rather than in `shared`. Just to be clear, I _don't_ want a per-baord error table anywhere.

thekenu commented on boards/DIM/Src/App/App_ErrorTable.c:
```diff
@@ -0,0 +1,216 @@
+    if (error_table->bms_stack_watermark_above_threshold_task1hz)
+    {
+        errors[num_errors].error_id = BMS_STACK_WATERMARK_ABOVE_THRESHOLD_TASK1HZ;
+        errors[num_errors].std_id = 104;
```
`cantools` will allow us to generate the correct std_id

thekenu commented on boards/DIM/Inc/App/App_ErrorTable.h:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Check if there is any critical errors in the given error table, and if
+ * necessary, write the critical errors into the given error array
+ * @note The array must hold up to NUM_ERRORS of elements
```
I'm not a big fan of how the user must manually make sure the array is large enough. Might consider generating a struct for the caller to use

```
struct ErrorArray {
  struct Error errors[NUM_ERRORS];
}; 
 uint32_t App_ErrorTable_HasCriticalError(struct ErrorTable* error_table, struct ErrorArray* error_array); 
```

garethellis0 replied:
Yeah, that's fine by me.

thekenu commented on boards/DIM/Src/main.c:
```diff
@@ -645,6 +649,7 @@ void RunTaskCanRx(void const *argument)
         Io_SharedCan_DequeueCanRxMessage(&message);
         Io_CanRx_UpdateRxTableWithMessage(
             App_DimWorld_GetCanRx(world), &message);
+        Io_ErrorTable_Update(App_DimWorld_GetErrorTable(world), &message);
```
Wasn't sure if `Io_ErrorTable_Update` should be called inside `Io_CanRx_UpdateRxTableWithMessage`. Left them separate for now.

thekenu commented on boards/shared/Src/Io/Io_SharedCan.c:
```diff
@@ -297,6 +298,10 @@ void Io_SharedCan_DequeueCanRxMessage(struct CanMsg *message)
         ;
 }
 
+// This extern is a hack for now. Need to rewrite
```
Similar concern as: https://github.com/UBCFormulaElectric/Consolidated-Firmware/pull/615/files#r426098243.

There is no nice way to separate Error Table and CAN TX for now though.

thekenu commented on boards/DIM/Test/Src/Test_StateMachine.cpp:
```diff
@@ -376,20 +380,43 @@ TEST_F(DimStateMachineTest, imd_led_control_in_drive_state)
 
+TEST_F(DimStateMachineTest, error_on_7_segment_displays)
+{
+    App_ErrorTable_SetBmsStackWaterMarkAboveThresholdTask1Hz(error_table);
```
I was hoping to replace this line with `App_CanRx_SettBmsStackWaterMarkAboveThresholdTask1Hz`, but since Error Table and Can RX are separate right now I can't do that. Not sure if that's a big deal.

thekenu commented on boards/DIM/Src/Io/Io_ErrorTable.c:
```diff
@@ -0,0 +1,103 @@
+#include "Io_ErrorTable.h"
+
+// This should be used in both CAN TX and CAN RX callbacks
+void Io_ErrorTable_Update(
```
If we're generating the code, I was considering generating `Io_ErrorTable_UpdateCanTx` and `Io_ErrorTable_UpdateRx`.

If we're generating code for BMS, the `Io_ErrorTable_UpdateCanTx` would handle only BMS errors and `Io_ErrorTable_UpdateRx` would handle errors from the rest of the boards.

thekenu commented on boards/DIM/Src/App/states/App_DriveState.c:
```diff
@@ -174,6 +177,95 @@ static void DriveStateRunOnTick100Hz(struct StateMachine *const state_machine)
+            {
+                case CANMSGS_BMS_CRITICAL_ERRORS_FRAME_ID:
+                {
+                    digits_to_display += 500;
```
The number after `+=` comes from DIM-9

garethellis0 commented on boards/DIM/Src/App/App_ErrorTable.c:
```diff
@@ -0,0 +1,216 @@
+    free(error_table);
+}
+
+uint32_t App_ErrorTable_HasNonCriticalError(struct ErrorTable* error_table, struct Error errors[])
```
Um. I'm a tad confused. This is a function in the `DIM` board (it should be shared?) and it also uses the stack watermark from the BMS for some reason?

thekenu replied:
It was just placed here as an example so I can check that the code compiles. And the stack watermark is just one of the example non-critical error in the error table.

garethellis0 commented on boards/DIM/Inc/App/App_ErrorTable.h:
```diff
@@ -0,0 +1,87 @@
+{
+    // CAN message ID
+    uint32_t std_id;
+    // Each error ID is unique in the error table
```
Do you mean in the `CAN` message? If the error ID is globally unique (ie. across all CAN messages) then there's no point to having the `std_id` here as well?

thekenu replied:
Good catch. This comment is wrong. The error IDs will be unique only on a per-board basis to fulfill:
![image](https://user-images.githubusercontent.com/16970019/82135690-63112280-97ba-11ea-9bb0-3f9d990a4a59.png)

thekenu replied:
Initially I thought about making the error IDs globally unique, but considering the fact that we only have 2 digits to display errors that would leave us with 99 errors in total across 5 PCBs. Which seemed too restrictive.

thekenu replied:
The per-board unique IDs are reflected in the enum at the top of this file

garethellis0 commented on boards/DIM/Inc/App/App_ErrorTable.h:
```diff
@@ -0,0 +1,87 @@
+ */
+uint32_t App_ErrorTable_HasCriticalError(struct ErrorTable* error_table, struct Error errors[]);
+
+void App_ErrorTable_SetBmsStackWaterMarkAboveThresholdTask1Hz(struct ErrorTable* error_table);
```
I'm confused why we're setting errors like this, but then define errors using the struct above? I originally proposed using the struct above so that we wouldn't need to have semantic understanding of what each error is.

thekenu replied:
Did you mean `struct Error`? I made these setters becuase I thought it's what you meant by

>// Setters for each possible error message ... we would determine these by regex match on the message ID's from the dbc

I must have misinterpreted what you meant? Maybe `Io_ErrorTable_Update` is what you wanted me to write?

thekenu commented:
I _think_ I know what you mean by raw CAN message level. If we go down that route, how do we figure out where each error came from and how would a user look up an error's semantic meaning given an error ID?

Also, I moved things under `shared/` and added `Shared` into the function names to indicate that things are meant to be shareable across boards.
